SpaceInvaders/src/Controllers/BulletsController.java

```diff
@@ -0,0 +1,127 @@
+            else battlefield[b.getPositionY() - 1][b.getPositionX()] = 6;
+        });
+        return battlefield;
+    }
+
+    public void MoveBullets(){
+        for(int i = 0; i < bullets.size(); i++){
+            if(bullets.get(i).getPositionY() < 17 && bullets.get(i).getEnemyBullet() == true) bullets.get(i).MoveBullet();
+            else if (bullets.get(i).getEnemyBullet() == true) bullets.remove(i);
+            else if(bullets.get(i).getPositionY() > 3 && bullets.get(i).getEnemyBullet() == false) bullets.get(i).MoveBullet();
+            else if (bullets.get(i).getEnemyBullet() == false) bullets.remove(i);
+        }
+    }
+
+    public ArrayList<EarthDefensiveBlock> CheckBulletsDefensiveBlockCollision(ArrayList<EarthDefensiveBlock> earthDefensiveBlocks){
+        for(int d = 0; d < bullets.size(); d++){
+            CheckDefensiveBlocksInBulletPosition(earthDefensiveBlocks, d);
+
+        }
+        return earthDefensiveBlocks;
+    }
+
+    private void CheckDefensiveBlocksInBulletPosition(ArrayList<EarthDefensiveBlock> earthDefensiveBlocks, int d) {
+        Boolean hit = false;
+        for(int c = 0; c < earthDefensiveBlocks.size(); c++){
+            hit = CheckIfAnyBricksInBlockWereHit(earthDefensiveBlocks, d, hit, c);
+            if(hit == true) break;
+        }
+    }
+
+    private Boolean CheckIfAnyBricksInBlockWereHit(ArrayList<EarthDefensiveBlock> earthDefensiveBlocks, int d, Boolean hit, int c) {
+        for(int i = 0; i < earthDefensiveBlocks.get(c).getDefensiveBlockBricks().size(); i++){
+            if(earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionX() == bullets.get(d).getPositionX() && earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionY() == bullets.get(d).getPositionY()){
+                hit = ComputeBrickBulletCollision(earthDefensiveBlocks, c, i, d);
+                break;
+            }
+        }
+        if(hit == false)
+            for(int i = 0; i < earthDefensiveBlocks.get(c).getDefensiveBlockBricks().size(); i++){
+                if(earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionX() == bullets.get(d).getPositionX() && earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionY() == bullets.get(d).getPositionY() - 1 && bullets.get(d).getEnemyBullet() == false){
+                    hit = ComputeBrickBulletCollision(earthDefensiveBlocks, c, i, d);
+                    break;
+                }
+                else if(earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionX() == bullets.get(d).getPositionX() && earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionY() == bullets.get(d).getPositionY() + 1 && bullets.get(d).getEnemyBullet() == true){
+                    hit = ComputeBrickBulletCollision(earthDefensiveBlocks, c, i, d);
+                    break;
+                }
+            }
+        return hit;
+    }

```

Comment:
Per sudėtingas metodas, reikia sukurti papildomu metodu atskirom funkcijom pvz:

    private Boolean CheckIfAnyBricksInBlockWereHit(ArrayList<EarthDefensiveBlock> earthDefensiveBlocks, int d, Boolean hit, int c) {
 for(int i = 0; i < earthDefensiveBlocks.get(c).getDefensiveBlockBricks().size(); i++){
            if(earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionX() == bullets.get(d).getPositionX() && earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionY() == bullets.get(d).getPositionY()){
                hit = ComputeBrickBulletCollision(earthDefensiveBlocks, c, i, d);
                break;
            }
        }
        if(hit == false)
            _for(int i = 0; i < earthDefensiveBlocks.get(c).getDefensiveBlockBricks().size(); i++){
                if(earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionX() == bullets.get(d).getPositionX() && earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionY() == bullets.get(d).getPositionY() - 1 && bullets.get(d).getEnemyBullet() == false){
                    hit = ComputeBrickBulletCollision(earthDefensiveBlocks, c, i, d);
                    break;
                }
                else if(earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionX() == bullets.get(d).getPositionX() && earthDefensiveBlocks.get(c).getDefensiveBlockBricks().get(i).getPositionY() == bullets.get(d).getPositionY() + 1 && bullets.get(d).getEnemyBullet() == true){
                    hit = ComputeBrickBulletCollision(earthDefensiveBlocks, c, i, d);
                    break;
                }_  Čia galima tiesiog MetodoPavadinimas()
            }
        return hit;
    }